### PR TITLE
interfaces/builtin: add i2c interface

### PR DESCRIPTION
**Pull request by bergotorino**
Add i2c interface which allows access to the registered i2c adapters in rw mode.

### CODE REVIEW

**Comment by lpotter** [comment]
Looks ok to me, although I would like to see a more generic and slightly more encompassing interface for things such as /dev/spidev

**Comment by morphis** [comment]
@lpotter SPI would go into an spi interface. Why would we make a generic interface allowing access to I2C and SPI together?

File: asserts/asserts.go [hunk 1]
```diff
@@ -86,7 +85,6 @@ var typeRegistry = map[string]*AssertionType{
 	ValidationType.Name:      ValidationType,
 	// no authority
 	DeviceSessionRequestType.Name: DeviceSessionRequestType,
-	SerialProofType.Name:          SerialProofType,
```
Comment (jdstrand) [inline asserts/asserts.go]: These SerialProof changes seem unrelated to this PR.

File: docs/interfaces.md [hunk 2]
```diff
@@ -138,6 +138,13 @@ owned by the user to read/write/lock.
 * Auto-Connect: yes on classic (traditional distributions), no otherwise
 * Transitional: yes
 
+### i2c
+
+Can access the registered i2c adapters in read-write mode. Suitable for
+talking to perpiherial devices.
```
Comment (jdstrand) [inline docs/interfaces.md]: Typo

File: interfaces/builtin/i2c.go [hunk 3]
```diff
@@ -17,8 +17,23 @@
...
 
-var (
-	ImportAssertionsFromSeed = importAssertionsFromSeed
+import (
+	"github.com/snapcore/snapd/interfaces"
 )
+
+const i2cConnectedPlugAppArmor = `
+/dev/i2c-[0-9]* rw,
+`
```
Comment (jdstrand) [inline interfaces/builtin/i2c.go]: I think `i2c` should probably be renamed to `i2c-control` if this stays as is.

However, I'm not sure of the direction of this PR. It seems like this should be more like hidraw and serial-port where these devices is the udev backend along with slot attributes provided by gadget yaml to expose the slots to consuming apps. Is there a reason you did not pursue this route?

CC: @niemeyer and @morphis

File: interfaces/builtin/i2c.go [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+ */
+
+package builtin
+
+import (
+	"github.com/snapcore/snapd/interfaces"
+)
+
+const i2cConnectedPlugAppArmor = `
+/dev/i2c-[0-9]* rw,
```
Comment (morphis) [inline interfaces/builtin/i2c.go]: Can we add 

/sys/class/i2c-dev/ r,
/sys/devices/**/i2c-dev/** rw,

to allow users to detect available nodes and get some information about them? i2cdetect uses this AFAIK.

File: interfaces/builtin/i2c.go (hunk 3 above)
Comment (morphis) [inline interfaces/builtin/i2c.go]: @jdstrand Discussed this with @niemeyer and we agreed that we will implement this the same way as the serial-port/hidraw interfaces. @bergotorino is working on that now.

**Comment by jdstrand** [comment]
Per @morphis and @niemeyer, this is going to be reworked. AIUI it is actively being worked on.

File: interfaces/builtin/i2c_control.go [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+	}
+	switch securitySystem {
+	case interfaces.SecurityAppArmor:
+		// No USB attributes for i2c so just go directly to
+		// considering the fixed node
+		cleanedPath := filepath.Clean(path)
+		return []byte(fmt.Sprintf("%s rw,\n", cleanedPath)), nil
+
+	case interfaces.SecurityUDev:
+		const udevRule string = `KERNEL=="%s", TAG+=snap_%s_%s`
```
Comment (zyga) [inline interfaces/builtin/i2c_control.go]: Why are you tagging it via udev?

File: interfaces/builtin/i2c_control.go [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+	"regexp"
+	"strings"
+)
+
+// The type for i2c control interface
+type I2cControlInterface struct{}
+
+// Getter for the name of the i2c-control interface
+func (iface *I2cControlInterface) Name() string {
+	return "i2c-control"
```
Comment (niemeyer) [inline interfaces/builtin/i2c_control.go]: Why "-control" here?  This feels more like "serial-port" or "hidraw" than "timezone-control" or "network-control".

File: interfaces/builtin/i2c_control.go [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+
+	// Does it have right type?
+	if iface.Name() != slot.Interface {
+		panic(fmt.Sprintf("slot is not of interface %q", iface))
+	}
+
+	// Creation of the slot of this type
+	// is allowed only by a gadget snap
+	if slot.Snap.Type != "gadget" {
+		return fmt.Errorf("i2c-control slots only allowed on gadget snaps")
```
Comment (niemeyer) [inline interfaces/builtin/i2c_control.go]: All messages need to be fixed if the interface name is changed.

File: interfaces/builtin/i2c_control.go [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+
+// Getter for the security snippet specific to the plug
+func (iface *I2cControlInterface) ConnectedPlugSnippet(plug *interfaces.Plug, slot *interfaces.Slot, securitySystem interfaces.SecuritySystem) ([]byte, error) {
+	path, pathOk := slot.Attrs["path"].(string)
+	if !pathOk {
+		return nil, nil
+	}
+	switch securitySystem {
+	case interfaces.SecurityAppArmor:
+		// No USB attributes for i2c so just go directly to
```
Comment (niemeyer) [inline interfaces/builtin/i2c_control.go]: If this is not USB, why are we even mentioning it here?  Makes me feel like I'm missing something. In other words, we don't need to mention all the attributes that it does _not_ have because they make no sense. If there's a gotcha, let's please make the comment more clear.

File: interfaces/builtin/i2c_control.go [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+		// considering the fixed node
+		cleanedPath := filepath.Clean(path)
+		return []byte(fmt.Sprintf("%s rw,\n", cleanedPath)), nil
+
+	case interfaces.SecurityUDev:
+		const udevRule string = `KERNEL="%s", TAG+="snap_%s_%s"`
```
Comment (niemeyer) [inline interfaces/builtin/i2c_control.go]: Why do we need the tagging here? Real question. It might make total sense.. I'm just not aware of how the security hooking is taking place via udev here.

File: snap/implicit.go [hunk 10]
```diff
@@ -62,6 +62,7 @@ var implicitClassicSlots = []string{
 	"browser-support",
 	"cups-control",
 	"gsettings",
+	"i2c",
```
Comment (niemeyer) [inline snap/implicit.go]: That doesn't make much sense I think. How can it be implicit if it requires a path attribute, and how can it be implicit in the core if the interface forces it to be in the gadget?

The interface name is also wrong per the current code.

File: interfaces/builtin/i2c_control.go (hunk 6 above)
Comment (bergotorino) [inline interfaces/builtin/i2c_control.go]: perhaps -bus then?

File: interfaces/builtin/i2c_control.go (hunk 9 above)
Comment (bergotorino) [inline interfaces/builtin/i2c_control.go]: We agreed on allowing the access to a particular i2c device nodes instead of the bus as a whole for a single snap. Here it is tagged to explicitly move the /dev/i2c-whatever to the control group for this snap to have access to. Does it make sense?

File: interfaces/builtin/i2c_control.go (hunk 6 above)
Comment (morphis) [inline interfaces/builtin/i2c_control.go]: This should go as i2c now that we have the same approach than we have with serial-port or hidraw.

File: interfaces/builtin/i2c_control.go (hunk 9 above)
Comment (morphis) [inline interfaces/builtin/i2c_control.go]: From what I remember there was a existing function in utils.go or somewhere else to generate "snap_%s_%s" so we don't have to repeat this here. Please use that function.

File: interfaces/builtin/i2c_control.go [hunk 11]
```diff
@@ -0,0 +1,133 @@
...
+		var udevSnippet bytes.Buffer
+		for appName := range plug.Apps {
+			rule := fmt.Sprintf(udevRule, strings.TrimPrefix(path, "/dev/"), plug.Snap.Name(), appName)
```
Comment (morphis) [inline interfaces/builtin/i2c_control.go]: Lets put /dev as constant somewhere.

File: interfaces/builtin/i2c_control.go (hunk 9 above)
Comment (bergotorino) [inline interfaces/builtin/i2c_control.go]: There is no such function. I think that I know to which one you are referring to but it works with USB devices: https://github.com/snapcore/snapd/blob/master/interfaces/builtin/utils.go#L74

File: docs/interfaces.md [hunk 12]
```diff
@@ -138,6 +138,13 @@ owned by the user to read/write/lock.
 * Auto-Connect: yes on classic (traditional distributions), no otherwise
 * Transitional: yes
 
+### i2c
+
+Can access the registered i2c adapters in read-write mode. Suitable for
+talking to peripherial devices.
+
+* Auto-Connect: yes
```
Comment (bergotorino) [inline docs/interfaces.md]: Good catch!

File: interfaces/builtin/i2c_control.go (hunk 9 above)
Comment (morphis) [inline interfaces/builtin/i2c_control.go]: Then lets create one. This is now the second or third place where this is repeated. something like InterfaceUdevSecurityTag(snapName, appName)

File: docs/interfaces.md [hunk 13]
```diff
@@ -138,6 +138,13 @@ owned by the user to read/write/lock.
 * Auto-Connect: yes on classic (traditional distributions), no otherwise
 * Transitional: yes
 
+### i2c
+
+Can access the registered i2c adapters in read-write mode. Suitable for
+talking to peripherial devices.
```
Comment (jdstrand) [inline docs/interfaces.md]: Typo: should be 'peripheral'; that said, I think it best to remove the whole 'Suitable for...' and just leave the 'Can access...' sentence.

File: interfaces/builtin/basedeclaration.go [hunk 14]
```diff
@@ -241,6 +241,11 @@ slots:
         - core
     deny-auto-connection:
       on-classic: false
+  i2c-control:
+    allow-installation:
+      slot-snap-type:
+        - gadget
```
Comment (jdstrand) [inline interfaces/builtin/basedeclaration.go]: This should be:
```
slot-snap-type:
  - core￼
  - gadget
```

File: interfaces/builtin/basedeclaration_test.go [hunk 15]
```diff
@@ -361,6 +361,7 @@ var (
 		"docker-support":  []string{"core"},
 		"gpio":            []string{"core", "gadget"},
 		"hidraw":          []string{"core", "gadget"},
+		"i2c-control":     []string{"gadget"},
```
Comment (jdstrand) [inline interfaces/builtin/basedeclaration_test.go]: This should be:
```
"i2c-control":     []string{"core", "gadget"},
```

File: interfaces/builtin/i2c_control.go (hunk 6 above)
Comment (jdstrand) [inline interfaces/builtin/i2c_control.go]: +1 for renaming this `i2c`

File: interfaces/builtin/i2c_control.go [hunk 16]
```diff
@@ -0,0 +1,133 @@
...
+}
+
+// Pattern to match allowed i2c device nodes. It is gonna be used to check the
+// validity of the path attributes in case the udev is not used for
+// identification
+var i2cControlDeviceNodePattern = regexp.MustCompile("^/dev/i2c-[0-9]+$")
+
+// Check validity of the defined slot
+func (iface *I2cControlInterface) SanitizeSlot(slot *interfaces.Slot) error {
+
```
Comment (jdstrand) [inline interfaces/builtin/i2c_control.go]: Please remove this whitespace.

File: interfaces/builtin/i2c_control.go (hunk 9 above)
Comment (jdstrand) [inline interfaces/builtin/i2c_control.go]: @bergotorino is correct that we want to use the udev backend here even though the device names aren't expected to change. This is needed if a snap plugs this interface and another interface that uses udev tagging so that the i2c devices end up in the device cgroup.

@morphis is right that we should use the existing code for the tagging. See the serial-port interface for examples.

**Comment by niemeyer** [comment]
Any news here?  It'd be good to have this in on today's release, which is very close now.

**Comment by morphis** [comment]
Lets not wait for this. If it's out of the Release then that is fine. I am not sure if Konrad still has time to work on this today. I am fine with marking this is not-critical.

**Comment by bergotorino** [comment]
I will work on this now.

File: interfaces/builtin/i2c_control.go (hunk 9 above)
Comment (bergotorino) [inline interfaces/builtin/i2c_control.go]: The problem though is that there is no function that could tag it the way we want. We need to write it just for this one purpose - that is the whole point. If there would be then it would be used already :)

File: interfaces/builtin/utils.go [hunk 17]
```diff
@@ -83,3 +83,15 @@ func udevUsbDeviceSnippet(subsystem string, usbVendor int64, usbProduct int64, k
 	udevSnippet.WriteString("\n")
 	return udevSnippet.Bytes()
 }
+
+// Function to support creation of udev TAG snippet for all the apps bound
+// to a given slot.
+func udevSecurityTagSnippet(deviceNode string, snapName string, apps map[string]*snap.AppInfo) []byte {
```
Comment (morphis) [inline interfaces/builtin/utils.go]: As said before, there is no need for this function. The only thing we should move into its own function is the creation of the "snap_%s_%s" string as that one happens at different places.

File: interfaces/builtin/utils.go [hunk 18]
```diff
@@ -83,3 +83,25 @@ func udevUsbDeviceSnippet(subsystem string, usbVendor int64, usbProduct int64, k
...
+// @param snapName is the name of the snap
+// @param appName is the name of the application
+// @return string "snap_<snap name>_<app name>"
+func udevTagString(snapName string, appName string) string {
+	return fmt.Sprintf(`snap_%s_%s`, snapName, appName)
+}
+
+// Function to support creation of udev TAG snippet for all the apps bound
+// to a given slot.
+func udevSecurityTagSnippet(deviceNode string, snapName string, apps map[string]*snap.AppInfo) []byte {
```
Comment (morphis) [inline interfaces/builtin/utils.go]: This one can be dropped now.

File: interfaces/builtin/utils.go [hunk 19]
```diff
@@ -83,3 +83,25 @@ func udevUsbDeviceSnippet(subsystem string, usbVendor int64, usbProduct int64, k
 	udevSnippet.WriteString("\n")
 	return udevSnippet.Bytes()
 }
+
+// Function to create an udev TAG, essentially the cgroup name for
+// the snap application.
+// @param snapName is the name of the snap
+// @param appName is the name of the application
+// @return string "snap_<snap name>_<app name>"
+func udevTagString(snapName string, appName string) string {
```
Comment (morphis) [inline interfaces/builtin/utils.go]: Better name would be something like udevSnapSecurityName
